pandas/_libs/src/klib/khash_python.h

```diff
@@ -330,6 +332,46 @@ Py_hash_t PANDAS_INLINE tupleobject_hash(PyTupleObject* key) {
 }
 
 
+// TODO: same thing for timedelta64 objects
+Py_hash_t np_datetime64_object_hash(PyObject* key) {
```

Comment:
Rather than check+cast below, you could do:
```suggestion
Py_hash_t np_datetime64_object_hash(PyDatetimeScalarObject* key) {
```
and move the cast to the caller.  That would ensure type correctness by at least getting a warning in C (I realize you guys may be ignoring warnings though).